internal/armada/server/authorization.go

```diff
@@ -0,0 +1,158 @@
+	reasons := make([]string, len(err.Reasons), len(err.Reasons))
+	for i, reason := range err.Reasons {
+		reasons[i] = fmt.Sprintf("%q %s", principalName, reason)
+	}
+	return strings.Join(reasons, ", ")
+}
+
+// checkReprioritizePerms checks if the principal embedded in the context has permission
+// to reprioritize all of the given jobs. This may require permissions for several queues.
+// If the principal has sufficient permissions, nil is returned. Otherwise an error is returned.
+func (server *SubmitServer) checkReprioritizePerms(ctx context.Context, jobs []*api.Job) error {
+
+	// Set of queues that has at least 1 of the given jobs
+	queues := make(map[string]bool)
+	for _, job := range jobs {
+		queues[job.Queue] = true
+	}
+
+	// Check permissions for all queues
+	// Record the names of queues for which the principal lacks permissions
+	var reasons []string
+	for queue := range queues {
+		_, err := server.checkQueuePermission(ctx, queue, false, permissions.ReprioritizeJobs, permissions.ReprioritizeAnyJobs)
+		var e *ErrNoPermission
+		if errors.As(err, &e) {
+			reasons = append(reasons, fmt.Sprintf("does not own queue %q and have %s permissions for it", queue, permissions.ReprioritizeJobs))
+		} else if err != nil {
+			return fmt.Errorf("[checkReprioritizePerms] error checking permissions: %w", err)
+		}
+	}
+
+	// If there is a queue for which the principal lacks permissions,
+	// return an error containing the names of those queues
+	if len(reasons) != 0 {
+		reasons = append(reasons, fmt.Sprintf("does not have %s permissions", permissions.ReprioritizeAnyJobs))
+		return &ErrNoPermission{
+			Principal: authorization.GetPrincipal(ctx),
+			Reasons:   reasons,
+		}
+	}
+
+	return nil
+}
+
+// checkQueuePermission checks if the principal embedded in the context has permission
+// to perform actions on the given queue. If the principal has sufficient permissions,
+// nil is returned. Otherwise an error is returned.
+func (server *SubmitServer) checkQueuePermission(
+	ctx context.Context,
+	queueName string,
+	attemptToCreate bool,
+	basicPermission permission.Permission,
+	allQueuesPermission permission.Permission) ([]string, error) {
+
+	// Load the queue into memory to check if the user is the owner of the queue
+	queue, err := server.queueRepository.GetQueue(queueName)
+	var e *repository.ErrQueueNotFound
+
+	// TODO Checking permissions shouldn't have side side effects.
+	// Hence, this function shouldn't automatically create queues.
+	// Further, we should consider removing the AutoCreateQueues option entirely.
+	// Since it leads to surprising behavior (e.g., creating a queue if the name is misspelled).
+	// Creating queues should always be an explicit decision.
+	// The less surprising behavior is to return ErrQueueNotFound (perhaps wrapped).
+	if errors.As(err, &e) && attemptToCreate && server.queueManagementConfig.AutoCreateQueues {
+		// TODO Is this correct? Shouldn't the relevant permission be permissions.CreateQueue?
```

Comment:
Good question

Reply:
This is the correct permission - for current usage. Not say it isn't a bit of an abuse of the roles.

However the use case is quite contrived and should probably be removed as stated in the above TODO comment - or be made more explicit. (i.e AutoCreateQueue permission, but that is pretty messy)

Reply:
We'd need to migrate existing users off this first

Reply:
All right